From 7aa9d32461d12add9b5fc2ca397e91cbabcde377 Mon Sep 17 00:00:00 2001
From: minaripenguin <minaripenguin@users.noreply.github.com>
Date: Sat, 26 Oct 2024 21:23:53 +0800
Subject: [PATCH] Launcher3: Fix crashes when navigating from launcher settings
 to home screen

*logs:

10-26 21:11:50.189 14667 14667 E AndroidRuntime: Process: com.android.launcher3, PID: 14667
10-26 21:11:50.189 14667 14667 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'android.graphics.drawable.Drawable$ConstantState com.android.launcher3.icons.FastBitmapDrawable.getConstantState()' on a null object reference
10-26 21:11:50.189 14667 14667 E AndroidRuntime: 	at com.android.launcher3.views.FloatingIconView.lambda$fetchIcon$3(FloatingIconView.java:570)
10-26 21:11:50.189 14667 14667 E AndroidRuntime: 	at com.android.launcher3.views.FloatingIconView$$ExternalSyntheticLambda3.get(D8$$SyntheticClass:0)
10-26 21:11:50.189 14667 14667 E AndroidRuntime: 	at com.android.launcher3.views.FloatingIconView.setOriginalDrawableBackground(FloatingIconView.java:404)
10-26 21:11:50.189 14667 14667 E AndroidRuntime: 	at com.android.launcher3.views.FloatingIconView.getFloatingIconView(FloatingIconView.java:635)
10-26 21:11:50.189 14667 14667 E AndroidRuntime: 	at com.android.quickstep.LauncherSwipeHandlerV2.createIconHomeAnimationFactory(LauncherSwipeHandlerV2.java:134)
10-26 21:11:50.189 14667 14667 E AndroidRuntime: 	at com.android.quickstep.LauncherSwipeHandlerV2.createHomeAnimationFactory(LauncherSwipeHandlerV2.java:128)
10-26 21:11:50.189 14667 14667 E AndroidRuntime: 	at com.android.quickstep.AbsSwipeUpHandler.animateToProgressInternal(AbsSwipeUpHandler.java:1617)

10-26 21:11:50.189 14667 14688 E AndroidRuntime: FATAL EXCEPTION: launcher-loader
10-26 21:11:50.189 14667 14688 E AndroidRuntime: Process: com.android.launcher3, PID: 14667
10-26 21:11:50.189 14667 14688 E AndroidRuntime: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.ComponentName android.content.Intent.getComponent()' on a null object reference
10-26 21:11:50.189 14667 14688 E AndroidRuntime: 	at android.content.pm.LauncherApps.resolveActivity(LauncherApps.java:975)
10-26 21:11:50.189 14667 14688 E AndroidRuntime: 	at com.android.launcher3.Utilities.getFullDrawable(Utilities.java:660)
10-26 21:11:50.189 14667 14688 E AndroidRuntime: 	at com.android.launcher3.views.FloatingIconView.getIconResult(FloatingIconView.java:319)
10-26 21:11:50.189 14667 14688 E AndroidRuntime: 	at com.android.launcher3.views.FloatingIconView.lambda$fetchIcon$4(FloatingIconView.java:585)

test: manual, swipe from launcher settings to homescreen

Signed-off-by: minaripenguin <minaripenguin@users.noreply.github.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 src/com/android/launcher3/Utilities.java              | 6 +++++-
 src/com/android/launcher3/views/FloatingIconView.java | 4 +++-
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/com/android/launcher3/Utilities.java b/src/com/android/launcher3/Utilities.java
index d20a247f8a..dba647f7aa 100644
--- a/src/com/android/launcher3/Utilities.java
+++ b/src/com/android/launcher3/Utilities.java
@@ -647,8 +647,12 @@ public final class Utilities {
                     ((PendingAddShortcutInfo) info).getActivityInfo(context);
             mainIcon = activityInfo.getFullResIcon(appState.getIconCache());
         } else if (info.itemType == LauncherSettings.Favorites.ITEM_TYPE_APPLICATION) {
+            android.content.Intent intent = info.getIntent();
+            if (intent == null) {
+                return null;
+            }
             LauncherActivityInfo activityInfo = context.getSystemService(LauncherApps.class)
-                    .resolveActivity(info.getIntent(), info.user);
+                    .resolveActivity(intent, info.user);
             if (activityInfo == null) {
                 return null;
             }
diff --git a/src/com/android/launcher3/views/FloatingIconView.java b/src/com/android/launcher3/views/FloatingIconView.java
index 5b3abc383a..88fc26e04f 100644
--- a/src/com/android/launcher3/views/FloatingIconView.java
+++ b/src/com/android/launcher3/views/FloatingIconView.java
@@ -571,7 +571,9 @@ public class FloatingIconView extends FrameLayout implements
             } else {
                 btvIcon = btv.getIcon();
                 // Clone when needed
-                btvDrawableSupplier = () -> btvIcon.getConstantState().newDrawable();
+                btvDrawableSupplier = () -> (btvIcon != null && btvIcon.getConstantState() != null)
+                        ? btvIcon.getConstantState().newDrawable()
+                        : null;
             }
         } else {
             btvIcon = null;
-- 
2.51.2.windows.1

